keps/sig-scheduling/5598-opportunistic-batching/README.md

```diff
@@ -0,0 +1,967 @@
+
+We are also working on gang scheduilng (in addition to other forms of multi-pod scheduling), which will give us a way to consider multiple pods at the same time. "Opportunistic batching" provides a starting point for these mechanisms by providing signatures and batching, both necessary foundational mechanisms, and including them initially in a simple way.
+
+Another change is the shift towards 1-pod-per-node in batch and ML environments. Many of these environments (among others) only attempt to run a single user pod on each node, along with a complement of daemon set pods. This simplifies our scheduling needs significantly, as it allows to reuse not only filtering, but also scoring results.
+
+### Goals
+
+ * Improve the performance of scheduling large jobs on large clusters where the constraints are simple.
+ * Begin building infrastructure to support gang scheduling and other "multi-pod" scheduling mechanisms.
+ * Ensure that the infrastructure we build is maintainable as we update, add and remove plugins.
+ * Provide improved performance with identical results to our current scheduler for targeted workloads in this release.
+ * Provide a path where we can expand batching to apply to most or all workloads over the next few releases.
+
+### Non-Goals
+
+ * We are not attempting to apply this optimization to all pods in this release. We will make the addition of batching transparent, but only applicable to a reduced set of workloads in v1.
+ * We are not adding gang scheduling of any kind in this KEP. This is purely a performance improvement without adding dependency on the Workload API [KEP-4671](https://github.com/kubernetes/enhancements/pull/5558), although we hope the work on this KEP will help us with gang scheduling as we build it.
+
+## Proposal
+
+See https://github.com/bwsalmon/kubernetes/pull/1 for a WIP version of the code (currently out of date). We discuss each of the added items: pod scheduling signature, batching mechanism and opportunistic batching in turn.
+
+### Pod scheduling signature
+
+The pod scheduling signature is used to determine if two pods are "the same" from a scheduling perspective. In specific, what this means is that any pod with the given signature will get the same scores / feasibility results from any arbitrary set of nodes. This is necessary for the cache to work, since we need to be able to reuse the previous work.
+
+Note that some pods will not have a signature, because the scoring uses not just pod and node attributes, but other pods in the system, global data about pod placement, etc. These nodes get a nil signature, and we fallback to the slow path.
```

Comment:
nit: probably, say the existing path, instead of the slow path so that we can clarify that path is just the same / not slower than today's scheduler.